crates/ruff_textwrap/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+use std::cmp;
+
+use ruff_newlines::StrExt;
+
+/// Indent each line by the given prefix.
+///
+/// # Examples
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent("First line.\nSecond line.\n", "  "),
+///            "  First line.\n  Second line.\n");
+/// ```
+///
+/// When indenting, trailing whitespace is stripped from the prefix.
+/// This means that empty lines remain empty afterwards:
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent("First line.\n\n\nSecond line.\n", "  "),
+///            "  First line.\n\n\n  Second line.\n");
+/// ```
+///
+/// Notice how `"\n\n\n"` remained as `"\n\n\n"`.
+///
+/// This feature is useful when you want to indent text and have a
+/// space between your prefix and the text. In this case, you _don't_
+/// want a trailing space on empty lines:
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent("foo = 123\n\nprint(foo)\n", "# "),
+///            "# foo = 123\n#\n# print(foo)\n");
+/// ```
+///
+/// Notice how `"\n\n"` became `"\n#\n"` instead of `"\n# \n"` which
+/// would have trailing whitespace.
+///
+/// Leading and trailing whitespace coming from the text itself is
+/// kept unchanged:
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent(" \t  Foo   ", "->"), "-> \t  Foo   ");
+/// ```
+pub fn indent(text: &str, prefix: &str) -> String {
+    let mut result = String::with_capacity(2 * text.len());
+    let trimmed_prefix = prefix.trim_end();
+    for line in text.universal_newlines() {
+        if line.trim().is_empty() {
+            result.push_str(trimmed_prefix);
+        } else {
+            result.push_str(prefix);
+        }
+        result.push_str(line.as_full_str());
+    }
+    result
+}
+
+/// Removes common leading whitespace from each line.
+///
+/// This function will look at each non-empty line and determine the
+/// maximum amount of whitespace that can be removed from all lines:
+///
+/// ```
+/// use ruff_textwrap::dedent;
+///
+/// assert_eq!(dedent("
+///     1st line
+///       2nd line
+///     3rd line
+/// "), "
+/// 1st line
+///   2nd line
+/// 3rd line
+/// ");
+/// ```
+pub fn dedent(text: &str) -> Cow<'_, str> {
+    // Find the minimum amount of leading whitespace on each line.
+    let mut prefix = None;
+    for line in text.universal_newlines() {
+        if let Some((index, _)) = line.char_indices().find(|(_, ch)| !ch.is_whitespace()) {
+            prefix = Some(cmp::min(prefix.unwrap_or(index), index));
+        }
+    }
+    let Some(prefix) = prefix else {
+        return Cow::Borrowed(text);
+    };
+
+    // Remove the common prefix from each line.
+    let mut result = String::with_capacity(text.len());
+    for line in text.universal_newlines() {
+        if line.trim().is_empty() {
+            result.push_str(line.newline_str());
+        } else {
+            result.push_str(&line.as_full_str()[prefix..]);
+        }
+    }
+
+    Cow::Owned(result)
```

Comment:
Unlike the source crate, this doesn't between mixed whitespace styles. For example, the original crate has this test, which I removed:

```rust
#[test]
#[rustfmt::skip]
fn dedent_mixed_whitespace() {
    let x = [
        "\tfoo",
        "  bar",
    ].join("\n");
    let y = [
        "\tfoo",
        "  bar",
    ].join("\n");
    assert_eq!(dedent(&x), y);
}
```

It's trickier, because you have to compare each prefix, rather than tracking the prefix _length_ alone. But Python doesn't support mixed indentation anyway, so I biased towards leaving it alone.

Reply:
Python does support mixed indentation, rust python doesn't. 

> Indentation is rejected as inconsistent if a source file mixes tabs and spaces in a way that makes the meaning dependent on the worth of a tab in spaces; a [TabError](https://docs.python.org/3/library/exceptions.html#TabError) is raised in that case. [[source](https://docs.python.org/3/reference/lexical_analysis.html#indentation)]

It took me a long time to understand what that means. My understanding (from looking at the CPython source is that) the column (respecting the tab width) and the character (counting tabs as 1) of two indentations must be the same. But `\t ` and ` \t` are equal indents. 

I assume what you mean by mixed whitespace styles is to use `\t` and `  ` which would not be equal because it's 1 character vs 2.

Reply:
Woah, I did not realize this -- I thought it was always a `TabError`.